stdlib/LibGit2/src/repository.jl

```diff
@@ -141,12 +141,29 @@ function (::Type{T})(repo::GitRepo, spec::AbstractString) where T<:GitObject
     @assert repo.ptr != C_NULL
     @check ccall((:git_revparse_single, libgit2), Cint,
                  (Ptr{Ptr{Cvoid}}, Ptr{Cvoid}, Cstring), obj_ptr_ptr, repo, spec)
+    obj_ptr = obj_ptr_ptr[]
     # check object is of correct type
-    if T != GitObject && T != GitUnknownObject # FIXME free result on error?
-        t = Consts.OBJECT(obj_ptr_ptr[])
-        t == Consts.OBJECT(T) || throw(GitError(Error.Object, Error.ERROR, "Expected object of type $T, received object of type $(objtype(t))"))
+    if T != GitObject && T != GitUnknownObject
+        t = Consts.OBJECT(obj_ptr)
+        if t != Consts.OBJECT(T)
+            if obj_ptr != C_NULL
+                # free result
+                if t ==  Consts.OBJ_COMMIT
```

Comment:
This if/else nest is not required. Libgit2 can do this internally if you call `git_object_free`

Reply:
That's much better, I'll do that.